lib/MusicBrainz/Server/Controller/Role/Merge.pm

```diff
@@ -194,29 +199,86 @@ role {
 
         my %entity_id = map { $_->id => $_ } @$entities;
 
-        my $new_id = $form->field('target')->value or die 'Coludnt figure out new_id';
+        my $new_id = $form->field('target')->value or die q{Couldn't figure out new_id};
         my $new = $entity_id{$new_id};
         my @old_ids = grep { $_ != $new_id } @{ $form->field('merging')->value };
 
         log_assertion { @old_ids >= 1 } 'Got at least 1 entity to merge';
 
         $c->model('MB')->with_transaction(sub {
             if ($params->edit_type) {
-                $self->_insert_edit(
-                    $c, $form,
-                    edit_type => $params->edit_type,
-                    new_entity => {
-                        id => $new->id,
-                        name => $new->name,
-                        $self->_extra_entity_data($c, $form, $new),
+                $self->edit_action($c,
+                    form        => $params->merge_form,
+                    type        => $params->edit_type,
+                    edit_args   => {
+                        new_entity => {
+                            id => $new->id,
+                            name => $new->name,
+                            $self->_extra_entity_data($c, $form, $new),
+                        },
+                        old_entities => [ map +{
+                            id => $entity_id{$_}->id,
+                            name => $entity_id{$_}->name,
+                            $self->_extra_entity_data($c, $form, $entity_id{$_}),
+                        }, @old_ids ],
+                        (map { $_->name => $_->value } $form->edit_fields),
+                        $self->_merge_parameters($c, $form, $entities),
+                    },
+                    on_creation => sub {
+                        my ($edit, $form) = @_;
+
+                        if ($c->namespace eq 'release' && $form->field('merge_rgs')->value) {
+                            my $edit_id = $edit->id;
+
+                            my $new_rg_id = $new->release_group->id;
+                            # We want to make sure we're merging only different RGs!
+                            my @old_entities = map +{
+                                id => $entity_id{$_}->release_group->id,
+                                name => $entity_id{$_}->release_group->name,
+                            },
+                            grep { $entity_id{$_}->release_group->id != $new_rg_id }
+                            @old_ids;
```

Comment:
I think you can simplify this slightly by throwing a `map { $entity_id{$_}->release_group }` in there. :)
```suggestion
                            my @old_entities = map +{ id => $_->id, name => $_->name },
                                grep { $_->id != $new_rg_id }
                                map { $entity_id{$_}->release_group }
                                @old_ids;
```